src/translate-adapter-handlers.ts

```diff
@@ -194,16 +204,61 @@ async function translateIoPackage(): Promise<void> {
 	if (content.common.messages) {
 		console.log("Translate Messages");
 		for (const message of content.common.messages) {
-			console.log(`   Message: ${message.title.en}`);
+			console.log(gray(`   Message: ${message.title.en}`));
 			await translateNotExisting(message.title);
 			await translateNotExisting(message.text);
-			await translateNotExisting(message.linkText);
+			// test first if there is a linkText - since it's not mandatory
+			if (message.linkText) await translateNotExisting(message.linkText);
 		}
 	}
 	await writeJson(ioPackage, content, { spaces: 4, EOL });
 	console.log(`Successfully updated ${path.relative(".", ioPackage)}`);
 }
 
+async function loopJSON(content: object): Promise<object> {
+	for (const [key, value] of Object.entries(content)) {
+		if (key === "i18n" && value === false) {
+			console.log(
+				"Info: i18n-switch is set to false; No translation of jsonConfig needed; Exiting.",
+			);
+			return content;
+		}
+		if (typeof value === "object") {
+			if (value.en) {
+				console.log(gray(`Translating: "${value.en}"`));
+				await translateNotExisting(value);
+			} else {
+				if (value.title || value.tooltip || value.label || value.text) {
+					const logText =
+						value.title ||
+						value.tooltip ||
+						value.label ||
+						value.text;
+					if (typeof logText === "string")
+						console.log(
+							gray(
+								`No english language-tag found for text "${logText}". You'll need to provide an english text to use automated translation.`,
```

Comment:
Could it be that the code ignores the case that i18n is set to tru but (as ideally the default!!) the standard i18n language files are used?

Honestly ... "inline texts in jsonconfig files" should not be a standard case because not compatible with Weblate. So preference is clearly to use i18n=true and use have a i18n/ directory with the translation files.

So having this convenience feature here might bring more devs to the point to have the translations in the jsonconfig which is not what we want ;-) But on the other side having it is good but then the logic needs to consider this.

In a "wish world" I would say the feature should extract the texts from the jsonconfig file to language files ... and allows to run it again an updated jsonconfig, identifying new elements and adding the texts to the language files and then run the translation as usual :-)

Reply:
I fully understand your "stategic doubts" after having heard of this strategy - but unfotunately I wasn't able read anything about this strategy in the documenation when moving over from old config to jsonConfig. Maybe it makes sense to drop a hint to it here (https://github.com/ioBroker/ioBroker.admin/blob/master/src-rx/src/components/JsonConfigComponent/SCHEMA.md). A question on this topic in german forum remained without any answer (https://forum.iobroker.net/topic/57801/adapter-korrekt-%C3%BCbersetzen-jsonconfig-wie).

But anyways:
> Could it be that the code ignores the case that i18n is set to tru but (as ideally the default!!) the standard i18n language files are used?

That's fully possible, since I'm not aware of these files. From the docs (and my research in other adapters) I only found this inline thing and considered it very inconveniant to use - since it don't gets translated automatecally. That's the reason for this code and the PR.

BTW - but slightly off topic: There are inline translations in blockly files as well - what about them? How to handle them correctly?

>Honestly ... "inline texts in jsonconfig files" should not be a standard case because not compatible with Weblate.

As already said: There's nothing in the docs on that. Neither in the above mentioned jsonConfig nor in adapter translation (https://github.com/ioBrokerTranslator/doc/#purpose). But enough of complaining! 
If you don't like the code feel free to drop it. If you like it - someone with full understanding of the strategy and use cases needs to extend it - since I don't have an full overview.

And to be honest as well: The weblate stuff, the how-to-use and it's benefits is currently poorly documentend and the hurdels to get an adapter there are f***** high. In my opinion these are most probably the more important reasons why developers don't use it. Using inline tanslations is much easier and quicker - even when it doesn't get translated automatecally.

> In a "wish world" I would say the feature should extract the texts from the jsonconfig file to language files ... and allows to run it again an updated jsonconfig, identifying new elements and adding the texts to the language files and then run the translation as usual :-)

I totally agree - but as already said: Someone with the full knowledge of the translation system needs to implement this else it ends up in the same garbage than my (first) attempt. Maybe this will be me - but than someone needs to explain the strategy and the full weblate infrastructure to me.

Reply:
Hi,

that this "strategic goal" is not mentioned in any documentation is kind of right. In fact it should/would look like https://github.com/PArns/ioBroker.netatmo/tree/master/admin ... you have the jsonConfig with i18n=true and "label" all the "english" texts (or in fact the "translation file keys). Additionally you have in admin the i18n directory with folders by language and a translations.json in it with the key -> translated text.

But the Schema document you linked exactly also had this option listed! https://github.com/ioBroker/ioBroker.admin/blob/master/src-rx/src/components/JsonConfigComponent/SCHEMA.md#i18n ( I only have now switched #1 and #2)

The Forum thread I never was aware of :-( sorry. Updated Schema and thread.

Blockly: They need to be in the files when I knwo it correctly, so Blockly was not really thought about so far how to "automate" them.

Weblate: Getting an adapter into Weblate is a hurdle? I do not think so honestly :-) If you use the language files then it is very easy and just needs the creation of a ticket. Then it is added and you get PRs... I did that for maaany adapters already :-)

Ok, but now enough strategic discussion: I love to accept the code. The only check I would add it "i18n=true and label is just a text and there is a i18n folder in the same directory then the jsonConfig.json consider it for now as ok and not show the errr log line?

Reply:
> Weblate: Getting an adapter into Weblate is a hurdle? I do not think so honestly :-) If you use the language files then it is very easy and just needs the creation of a ticket. Then it is added and you get PRs... I did that for maaany adapters already :-)

In my opinion it's a hurdle. I'm one of those guys who think twice before spreading around their google or github credentials on foreign websites - and creating a new account is effort I avoid when it's not fully clear what the benefit is. 
- So the login is the first hurdle to me. The only thing I know at this point in time: Weblate is some translating system ioBroker uses. It obviously has the trust of ioBroker core developers, but not nessecarily mine. I don't know how it works in detail and how much effort it would raise to me to use it. 
If I thow all this over board I get to the
- Second hurdle: The needed PR to get my adapter into it. If I open it - I'm stuck with my work for an unknown period of time (since I don't know when PR gets approved). I'm aware that (in general) it's a good thing to use it - but I'm currently sitting here and want to get my shit done. But that blocks me. So I tend to skip it and use the second best option, which allows me to progress.

> Ok, but now enough strategic discussion: I love to accept the code. The only check I would add it "i18n=true and label is just a text and there is a i18n folder in the same directory then the jsonConfig.json consider it for now as ok and not show the errr log line?

Surely I'm able to extend the code in that way, but that does not sound like the final solution to me, is it? Esp. aber having a look into the netatmo adapter
So if it's not - why not doing it right and solve the root cause? 

If I understood you correctly the root cause why jsonConfig doesn't get translated automatically is that the words/keys are missing in ./i18n/en/translations.json to get it translated there? But doesn't that mean that the "adminWords2Lang"-function does not grab them from jsonConfig? If this is the case the right way to solve this would be to 
- fully drop the message mentioned by you
- extend the adminWord2Lang code to grab labels, tooltips and titles from jsonConfig

Right?

Reply:
What do you think about the change to support tranlations.json files?

This will result in output like this:
![grafik](https://user-images.githubusercontent.com/5459426/195076577-82c78646-52af-43ec-80f0-d89c09a8d251.png)